crates/state/src/global_state.rs

```diff
@@ -24,264 +28,137 @@ impl GlobalState {
-        self.read_and_decode::<AccountMut>(&AccountMut::key(account_addr))
-            .await
-            .map(|res| res.map(|data| data.counter))
-    }
-
-    pub async fn set_account_balance(
-        &mut self,
-        account_addr: &Address,
-        balance: u64,
-    ) -> Result<()> {
-        self.replace(&AccountMut::key(account_addr), |mut data: AccountMut| {
-            data.balance = balance;
-            data
-        })
-        .await
-    }
-
-    pub async fn set_account_counter(
-        &mut self,
-        account_addr: &Address,
-        counter: u64,
-    ) -> Result<()> {
-        self.replace(&AccountMut::key(account_addr), |mut data: AccountMut| {
-            data.counter = counter;
-            data
-        })
-        .await
-    }
-
-    // TEMPLATE SECTIONS
-    // -----------------
-
-    pub async fn template_sections(&self, template_addr: &TemplateAddr) -> Result<Sections> {
-        let mut core_sections: Sections = self
-            .read_and_decode(&keys::template_core(template_addr))
-            .await?
-            .unwrap();
-        let noncore_sections: Sections = self
-            .read_and_decode(&keys::template_noncore(template_addr))
-            .await?
-            .unwrap();
-
-        for s in noncore_sections.iter().cloned() {
-            core_sections.insert(s);
-        }
-        Ok(core_sections)
-    }
-
-    pub async fn set_template_core_sections(
-        &mut self,
-        template_addr: &TemplateAddr,
-        sections: &Sections,
-    ) -> Result<()> {
-        self.encode_and_write(sections, &keys::template_core(template_addr))
-            .await;
-        Ok(())
+    pub fn in_memory() -> Self {
+        Self::new(":memory:")
     }
 
-    pub async fn set_template_noncore_sections(
-        &mut self,
-        template_addr: &TemplateAddr,
-        sections: &Sections,
-    ) -> Result<()> {
-        self.encode_and_write(sections, &keys::template_noncore(template_addr))
-            .await;
-        Ok(())
+    /// Returns a mutable reference to the [`Option<Layer>`] that controls the
+    /// historical query parameter used in *all* read operations. When set to
+    /// [`None`], the most recent value is always read; when set to
+    /// [`Some<some_layer>`], only the value present at the time of `some_layer`
+    /// is read.
+    pub fn historical_query(&mut self) -> &mut Option<Layer> {
+        &mut self.historical_query_context
     }
 
     // VERSIONING
     // ----------
 
-    pub async fn checkpoint(&mut self) -> Result<()> {
-        self.storage.checkpoint().await?;
+    /// Saves dirty changes in preparation of [`GlobalState::commit`]. After
+    /// saving, changes are frozen and can't be removed from the current layer.
+    ///
+    /// This might return a [`StorageError::KeyCollision`] depending on the
+    /// content of the dirty changes, so beware.
+    pub fn checkpoint(&mut self) -> Result<()> {
+        self.block_on(self.storage().checkpoint())?;
```

Comment:
not critical but I'd personally decompose into: (to reduce some cognitive load)

```rust
let storage = self.storage();
self.block_on(storage.checkpoint())?
```

Reply:
I'd prefer to leave these as simple one-liners if that's okay with you! I don't think there's much added value in `let storage = self.storage()`.

Reply:
I've kept this one open and marked as Resolved all the others.

Reply:
ok, it's up to you (it's a personal taste matter)